File: trace/ringbuf.go
/*
 * SPDX-License-Identifier: AGPL-3.0-only
 * Copyright (c) 2022-2025, daeuniverse Organization <dae@v2raya.org>
 */

package trace

import (
	"fmt"
	"math"
	"math/bits"
	"strconv"
	"strings"
)

const (
	DefaultRingbufSize      = "64MiB"
	minRingbufSizeBytes     = 4 << 10
	ringbufSizeAlignment    = 4 << 10
	defaultRingbufSizeBytes = 64 << 20
)

var ringbufSizeSuffixes = []struct {
	suffix     string
	multiplier uint64
}{
	{suffix: "gib", multiplier: 1 << 30},
	{suffix: "gb", multiplier: 1 << 30},
	{suffix: "g", multiplier: 1 << 30},
	{suffix: "mib", multiplier: 1 << 20},
	{suffix: "mb", multiplier: 1 << 20},
	{suffix: "m", multiplier: 1 << 20},
	{suffix: "kib", multiplier: 1 << 10},
	{suffix: "kb", multiplier: 1 << 10},
	{suffix: "k", multiplier: 1 << 10},
	{suffix: "b", multiplier: 1},
}

func ParseRingbufSizeBytes(value string) (uint32, error) {
	raw := strings.TrimSpace(value)
	if raw == "" {
		raw = DefaultRingbufSize
	}

	size, err := parseBinaryByteSize(raw)
	if err != nil {
		return 0, err
	}
	if size < minRingbufSizeBytes {
		return 0, fmt.Errorf("ring buffer size %q is too small; expect at least %d bytes", raw, minRingbufSizeBytes)
	}
	if size%ringbufSizeAlignment != 0 {
		return 0, fmt.Errorf("ring buffer size %q must be aligned to %d bytes", raw, ringbufSizeAlignment)
	}
	if bits.OnesCount64(size) != 1 {
		return 0, fmt.Errorf("ring buffer size %q must be a power of two", raw)
	}
	if size > math.MaxUint32 {
		return 0, fmt.Errorf("ring buffer size %q exceeds uint32 map limit", raw)
	}
	return uint32(size), nil
}

func DefaultRingbufSizeBytes() uint32 {
	return defaultRingbufSizeBytes
}

func parseBinaryByteSize(value string) (uint64, error) {
	normalized := strings.ToLower(strings.TrimSpace(value))
	if normalized == "" {
		return 0, fmt.Errorf("ring buffer size cannot be empty")
	}

	multiplier := uint64(1)
	numberPart := normalized
	for _, suffix := range ringbufSizeSuffixes {
		if strings.HasSuffix(normalized, suffix.suffix) {
			multiplier = suffix.multiplier
			numberPart = strings.TrimSpace(normalized[:len(normalized)-len(suffix.suffix)])
			break
		}
	}
	if numberPart == "" {
		return 0, fmt.Errorf("ring buffer size %q is missing its numeric value", value)
	}

	base, err := strconv.ParseUint(numberPart, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid ring buffer size %q: %w", value, err)
	}
	if base > math.MaxUint64/multiplier {
		return 0, fmt.Errorf("ring buffer size %q overflows uint64", value)
	}
	return base * multiplier, nil
}

File: rust/crates/dae-trace/src/ringbuf.rs
use std::fmt;

pub const DEFAULT_RINGBUF_SIZE: &str = "64MiB";
pub const MIN_RINGBUF_SIZE_BYTES: u64 = 4 << 10;
pub const RINGBUF_SIZE_ALIGNMENT: u64 = 4 << 10;
pub const DEFAULT_RINGBUF_SIZE_BYTES: u64 = 64 << 20;

const SUFFIXES: &[(&str, u64)] = &[
    ("gib", 1 << 30),
    ("gb", 1 << 30),
    ("g", 1 << 30),
    ("mib", 1 << 20),
    ("mb", 1 << 20),
    ("m", 1 << 20),
    ("kib", 1 << 10),
    ("kb", 1 << 10),
    ("k", 1 << 10),
    ("b", 1),
];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RingbufSizeError {
    message: String,
}

impl RingbufSizeError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for RingbufSizeError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.message)
    }
}

impl std::error::Error for RingbufSizeError {}

pub fn default_ringbuf_size_bytes() -> u64 {
    DEFAULT_RINGBUF_SIZE_BYTES
}

pub fn parse_ringbuf_size_bytes(value: &str) -> Result<u64, RingbufSizeError> {
    let raw = if value.trim().is_empty() {
        DEFAULT_RINGBUF_SIZE
    } else {
        value.trim()
    };

    let size = parse_binary_byte_size(raw)?;
    if size < MIN_RINGBUF_SIZE_BYTES {
        return Err(RingbufSizeError::new(format!(
            "ring buffer size {raw:?} is too small; expect at least {MIN_RINGBUF_SIZE_BYTES} bytes"
        )));
    }
    if size % RINGBUF_SIZE_ALIGNMENT != 0 {
        return Err(RingbufSizeError::new(format!(
            "ring buffer size {raw:?} must be aligned to {RINGBUF_SIZE_ALIGNMENT} bytes"
        )));
    }
    if !size.is_power_of_two() {
        return Err(RingbufSizeError::new(format!(
            "ring buffer size {raw:?} must be a power of two"
        )));
    }
    if size > u32::MAX as u64 {
        return Err(RingbufSizeError::new(format!(
            "ring buffer size {raw:?} exceeds uint32 map limit"
        )));
    }
    Ok(size)
}

fn parse_binary_byte_size(value: &str) -> Result<u64, RingbufSizeError> {
    let normalized = value.trim().to_ascii_lowercase();
    if normalized.is_empty() {
        return Err(RingbufSizeError::new("ring buffer size cannot be empty"));
    }

    let mut multiplier = 1_u64;
    let mut number_part = normalized.as_str();
    for (suffix, suffix_multiplier) in SUFFIXES {
        if normalized.ends_with(suffix) {
            multiplier = *suffix_multiplier;
            number_part = normalized[..normalized.len() - suffix.len()].trim();
            break;
        }
    }

    if number_part.is_empty() {
        return Err(RingbufSizeError::new(format!(
            "ring buffer size {value:?} is missing its numeric value"
        )));
    }

    let base = number_part.parse::<u64>().map_err(|err| {
        let kind = if err.kind() == &std::num::IntErrorKind::PosOverflow {
            "value out of range"
        } else {
            "invalid syntax"
        };
        RingbufSizeError::new(format!(
            "invalid ring buffer size {value:?}: strconv.ParseUint: parsing {number_part:?}: {kind}"
        ))
    })?;

    base.checked_mul(multiplier).ok_or_else(|| {
        RingbufSizeError::new(format!("ring buffer size {value:?} overflows uint64"))
    })
}
